pkg/spdk/engine.go

```diff
@@ -1229,6 +1346,415 @@ func (e *Engine) validateAndUpdateReplicaNvme(replicaName string, bdev *spdktype
 	return e.ReplicaStatusMap[replicaName].Mode, nil
 }
 
+// This method performs an online volume expansion for the Longhorn Engine using SPDK. It:
+// Expands underlying replica logical volumes (lvol)
+// Recreates the SPDK RAID bdev
+// Suspends and resumes frontend I/O as needed
+// Ensures cleanup and status update on failure
+func (e *Engine) Expand(spdkClient *spdkclient.Client, size uint64) (retErr error) {
+	e.Lock()
+	defer e.Unlock()
+
+	e.log.Info("Expanding engine")
+
+	// fetch all replica clients
+	replicaClients, err := e.getReplicaClients()
+	if err != nil {
+		retErr = err
+		return retErr
+	}
+	defer e.closeRplicaClients(replicaClients)
+
+	// startExpansion checks if the expansion can be started, and marks the expansion as in progress
+	requireExpansion, err := e.startExpansion(size, replicaClients)
+	if err != nil {
+		return errors.Wrap(err, "startExpansion failed")
+	}
+	if !requireExpansion {
+		e.log.Info("No need to expand engine")
+		e.SpecSize = size
+		return nil
+	}
+
+	var (
+		expanded  = false
+		engineErr error
+	)
+
+	// engineErr will be set when the engine failed to do any non-recoverable operations.
+	defer func() {
+		if retErr != nil {
+			e.log.WithError(engineErr).Errorf("Engine %s failed to expand", e.Name)
+			e.lastExpansionError = errors.Wrap(retErr, "engine failed to expand expansion").Error()
+			e.lastExpansionFailedAt = time.Now().UTC().Format(time.RFC3339Nano)
+		}
+
+		if engineErr != nil {
+			if e.State != types.InstanceStateError {
+				e.State = types.InstanceStateError
+			}
+			e.ErrorMsg = engineErr.Error()
+			e.log.WithError(engineErr).Errorf("Engine %s under non-recoverable operation during expansion", e.Name)
+
+			if errUpdateLogger := e.log.UpdateLogger(logrus.Fields{
```

Comment:
nit: See this logger update logic repeatedly but the only variable part is `logrus.Fields`. Suggest abstracting this a util function/method instead.